## Data Profiling and Bias Analysis

This notebook uses the [Pandas Profiling library](https://github.com/pandas-profiling/pandas-profiling) to create documentation summarizing the columns in the data. The output provides insight on the number of null values, the distribution of the column (or frequencies if categorical data). This allows us to quickly understand our data and possibly highlight any outliers or unexpected values.

The second part of this notebook explores our missing values to confirm whether there is a risk of bias based on missing data for particular Boroughs.

We import the required libraries

In [1]:
import pandas as pd
import numpy as np

In [2]:
from pandas_profiling import ProfileReport

To be able to view all column names, we're increasing the number of columns to be displayed on our screen.

In [3]:
pd.set_option('display.max_columns', None)

### Reading in Processed Data

We read in the processed NYC data.

In [4]:
df_nyc = pd.read_csv("processed/nyc_processed_2017_to_2019_20210225.csv")
df_nyc.head()

census_tract_GEOID  total-households  total-renter-occupied-households  \
0         36085013204              1790                               422   
1         36085013800              2369                               441   
2         36085014700              1341                               241   
3         36085019700               712                                99   
4         36085020804              1988                               179   

   total-owner-occupied-households  total-owner-occupied-households-mortgage  \
0                             1368                                       798   
1                             1928                                      1030   
2                             1100                                       707   
3                              613                                       366   
4                             1809                                      1270   

   median-gross-rent  median-household-income  median-property-value  \
0               1411                    84866                 493400   
1               1185                    82361                 561100   
2               1220                    84310                 463800   
3               1384                    98167                 449800   
4               1175                    95417                 602200   

   median-monthly-housing-cost  pct-white  pct-af-am  pct-hispanic  \
0                         2396       95.4        0.0           9.6   
1                         2404       98.8        0.5           6.2   
2                         2406       94.9        2.2           4.5   
3                         2297       95.4        1.0           8.3   
4                         2820       92.6        0.0           3.4   

   pct-am-indian  pct-asian  pct-nh-pi  pct-multiple  pct-other  \
0            0.0        1.3        0.0           0.6        2.7   
1            0.0        0.0        0.0           0.7        0.0   
2            0.0        3.0        0.0           0.0        0.0   
3            0.0        1.3        0.0           0.0        2.4   
4            0.0        4.5        0.0           1.0        2.0   

   pct-below-poverty-level  households-children  single-parent-household  \
0               -888888888                  382                       27   
1               -888888888                  585                       67   
2               -888888888                  366                       30   
3               -888888888                  219                       18   
4               -888888888                  535                       29   

   older-adult-alone  level-of-education  immigrant-status  english-fluency  \
0                260                3443               244             1172   
1                313                4553               117             1388   
2                155                2492                59              442   
3                 74                1283                49              158   
4                147                4142                79             1027   

   drive-to-work  public-transport-to-work  vacant-properties  \
0           1307                       672                115   
1           1581                       681                 92   
2           1023                       435                163   
3            622                       162                 18   
4           1754                       556                 67   

   live-in-mobile-home  pct-renter-occupied  pct-owner-occupied  \
0                    0            23.575419           76.424581   
1                    0            18.615450           81.384550   
2                    0            17.971663           82.028337   
3                    0            13.904494           86.095506   
4                    8             9.004024           90.995976   

   pct-owner-occupied-mortgage  pct-owner-occupied-without-mortgage  \
0    

As well as the Hillsborough data.

In [5]:
df_hills = pd.read_csv("processed/hillsborough_fl_processed_2017_to_2019_20210225.csv")
df_hills.head()

census_tract_GEOID  total-households  total-renter-occupied-households  \
0         12057010103              1454                               283   
1         12057011006              1861                               244   
2         12057011108               681                                41   
3         12057011203              1403                               552   
4         12057011206              1263                               676   

   total-owner-occupied-households  total-owner-occupied-households-mortgage  \
0                             1171                                       672   
1                             1617                                      1072   
2                              640                                       262   
3                              851                                       578   
4                              587                                       343   

   median-gross-rent  median-household-income  median-property-value  \
0                831                    63611                 153500   
1               1349                    66815                 194200   
2                497                    49821                 146000   
3                967                    72716                 235400   
4                750                    33329                 132500   

   median-monthly-housing-cost  pct-white  pct-af-am  pct-hispanic  \
0                         1226       92.1        1.6          11.8   
1                         1423       93.0        1.4           8.9   
2                         1379       91.0        1.9          16.4   
3                         1484       81.3       10.3          21.7   
4                         1168       78.9       15.5          16.2   

   pct-am-indian  pct-asian  pct-nh-pi  pct-multiple  pct-other  \
0            0.0        2.0        0.0           0.0        4.3   
1            0.0        1.0        0.0           3.4        1.2   
2            0.0        3.4        0.0           0.0        3.7   
3            0.0        0.6        2.9           1.3        3.6   
4            0.0        1.6        0.0           1.7        2.3   

   pct-below-poverty-level  households-children  single-parent-household  \
0               -888888888                  383                       57   
1               -888888888                  569                       60   
2               -888888888                   79                       14   
3               -888888888                  392                      102   
4               -888888888                  205                       70   

   older-adult-alone  level-of-education  immigrant-status  english-fluency  \
0                 92                3033               104              523   
1                222                3598                65              622   
2                205                1208               143              371   
3                 76                2683               267              808   
4                190                2057               282              744   

   drive-to-work  public-transport-to-work  vacant-properties  \
0           1711                         0                 95   
1           2227                         7                173   
2            570                         0                 76   
3           1579                         0                230   
4            823                        90                148   

   live-in-mobile-home  pct-renter-occupied  pct-owner-occupied  \
0                  801            19.463549           80.536451   
1                  197            13.111231           86.888769   
2                  357             6.020558           93.979442   
3                    0            39.344262           60.655738   
4                  361            53.523357           46.476643   

   pct-owner-occupied-mortgage  pct-owner-occupied-without-mortgage  \
0    

### Creating Profiles

We create the profiles by passing our data and specifying the title of the document. Pandas profiling can provide the following:

- Type inference: detect the types of columns in a dataframe.
- Essentials: type, unique values, missing values
- Quantile statistics like minimum value, Q1, median, Q3, maximum, range, interquartile range
- Descriptive statistics like mean, mode, standard deviation, sum, median absolute deviation, coefficient of variation, kurtosis, skewness
- Most frequent values
- Histogram
- Correlations highlighting of highly correlated variables, Spearman, Pearson and Kendall matrices
- Missing values matrix, count, heatmap and dendrogram of missing values
- Text analysis learn about categories (Uppercase, Space), scripts (Latin, Cyrillic) and blocks (ASCII) of text data.
- File and Image analysis extract file sizes, creation dates and dimensions and scan for truncated images or those containing EXIF information.

We are disabling expensive computations (such as correlations and dynamic binning) by passing `minimal=True`.

In [6]:
profile = ProfileReport(df_nyc, title='NYC Processed Data', minimal=True)

We are saving our profile as a HTML file.

In [ ]:
profile.to_file("NYC Processed Data.html")

We repeat for Hillsborough.

In [8]:
profile = ProfileReport(df_hills, title='Hillsborough Processed Data', minimal=True)

In [ ]:
profile.to_file("Hillsborough Processed Data.html")

### Futher Exploration

From the profiles, we can see that there are a few rows with null values and many placeholder values such as '666666666', '-666666666' and '-999999999'.

Focusing on NYC, we can quickly check if any rows include null values and how these are distributed across the boroughs. Using insight from [census.gov](https://www.census.gov/quickfacts/fact/table/newyorkcitynewyork,bronxcountybronxboroughnewyork,kingscountybrooklynboroughnewyork,newyorkcountymanhattanboroughnewyork,queenscountyqueensboroughnewyork,richmondcountystatenislandboroughnewyork/HSD410219#HSD410219) we can the proportion of housing units are as follow:
- Bronx: 16% (503,829)
- Brooklyn: 30% (958,567)
- Manhattan: 24% (759,460)
- Queens: 25% (778,932)
- Staten Island: 5% (166,246)

In [11]:
df_nyc[df_nyc.isna().any(axis=1)].groupby(['county']).count()[['census_tract_GEOID']]

census_tract_GEOID
county                           
Bronx                           9
Brooklyn                       27
Manhattan                      14
Queens                         47
Staten Island                   4

Queens has a disproportionate number of null values (47% compared to its household percentage of 25%).

Similar, we can look at some of these placeholder values.

In [12]:
df_nyc[df_nyc['median-house-age'] > 666666666].groupby(['county']).count()[['census_tract_GEOID']]

census_tract_GEOID
county                           
Bronx                           6
Brooklyn                       13
Manhattan                       9
Queens                         25
Staten Island                   3

In [13]:
df_nyc[df_nyc['pct-non-white'] > 666666666].groupby(['county']).count()[['census_tract_GEOID']]

census_tract_GEOID
county                           
Bronx                           5
Brooklyn                       11
Manhattan                       8
Queens                         21
Staten Island                   2

Queens has a disproportionate number of placeholder values for both `median-house-age` and `pct-non-white` (45% compared to its household percentage of 25%).

There are lots of scenarios where there are negative numbers (the place holder values).

In [14]:
cols = list(df_nyc.columns)[:-2]

In [15]:
df_nyc[(df_nyc[cols] < 0).any(axis=1)].groupby(['county']).count()[['census_tract_GEOID']]

census_tract_GEOID
county                           
Bronx                         339
Brooklyn                      761
Manhattan                     288
Queens                        669
Staten Island                 110

Brooklyn and Queens both have slightly elevated proportions for negative number placeholders (35% and 31% compared to their household percentages of 30% and 25% respectively). Meanwhile, Manhattan is substantially lower (13% compared to its household percentage of 24%).

### Raw NYC Evictions Data

We read in the raw NYC evictions data and view the first few rows.

In [9]:
df_raw_nyc = pd.read_csv("raw/nyc_evictions_geocoded.csv")
df_raw_nyc.head()

COURT_INDEX_NUMBER  DOCKET_NUMBER     EVICTION_ADDRESS EVICTION_APT_NUM  \
0           11371/16          33621    454 EAST 105TH ST              05C   
1          N77151/18         111600    601 WEST 189TH ST               4G   
2           68097/19          18347        2607 AVENUE O              5-B   
3           64883/18          25398   726 WILLOUGHBY AVE         BASEMENT   
4           95253/18          94512  945 SARATOGA AVENUE                1   

  EXECUTED_DATE MARSHAL_FIRST_NAME MARSHAL_LAST_NAME  \
0      12/12/18              Bruce              Kemp   
1        4/3/19            Darlene            Barone   
2       8/26/19             George       Essock, Jr.   
3       7/18/18               Gary              Rose   
4        9/4/19              Henry             Daley   

  RESIDENTIAL_COMMERCIAL_IND    BOROUGH  EVICTION_ZIP      Address CLEANED  \
0                Residential  MANHATTAN         10029    454 EAST 105TH ST   
1                Residential  MANHATTAN         10040    601 WEST 189TH ST   
2                Residential   BROOKLYN         11210        2607 AVENUE O   
3                Residential   BROOKLYN         11206   726 WILLOUGHBY AVE   
4                Residential   BROOKLYN         11212  945 SARATOGA AVENUE   

  STATE                             input_address match_indicator match_type  \
0    NY   454 EAST 105TH ST, MANHATTAN, NY, 10029           Match  Non_Exact   
1    NY   601 WEST 189TH ST, MANHATTAN, NY, 10040           Match  Non_Exact   
2    NY        2607 AVENUE O, BROOKLYN, NY, 11210           Match      Exact   
3    NY   726 WILLOUGHBY AVE, BROOKLYN, NY, 11206           Match      Exact   
4    NY  945 SARATOGA AVENUE, BROOKLYN, NY, 11212           Match      Exact   

                           matched_address              lon_lat  \
0      454 E 105TH ST, NEW YORK, NY, 10029  -73.93865,40.788143   
1      601 W 189TH ST, NEW YORK, NY, 10040  -73.930176,40.85436   
2          2607 AVE O, BROOKLYN, NY, 11210   -73.94823,40.61415   
3  726 WILLOUGHBY AVE, BROOKLYN, NY, 11206   -73.93993,40.69493   
4    945 SARATOGA AVE, BROOKLYN, NY, 11212   -73.91456,40.65798   

   tiger_line_id side  state_code  county_code  tract_code  block_code  
0    640857964.0    R        36.0         61.0     16200.0      4002.0  
1     59659038.0    R        36.0         61.0     27900.0      3000.0  
2     59090174.0    L        36.0         47.0     75600.0      2003.0  
3     59079108.0    R        36.0         47.0     28300.0      3000.0  
4     59084034.0    L        36.0         47.0     89600.0      3001.0

We could create another Pandas profiling document, however, we're going to quickly look at descriptive analytics using `describe()`.

In [10]:
df_raw_nyc.describe()

DOCKET_NUMBER  EVICTION_ZIP  tiger_line_id  state_code   county_code  \
count   60788.000000  60788.000000   5.545900e+04     55459.0  55459.000000   
mean   156367.932404  10801.659193   1.641214e+08        36.0     40.736346   
std    149694.482603    511.222055   2.146066e+08         0.0     29.667331   
min       103.000000      0.000000   5.907465e+07        36.0      5.000000   
25%     61667.750000  10456.000000   5.909886e+07        36.0      5.000000   
50%     91827.000000  10473.000000   5.986933e+07        36.0     47.000000   
75%    293285.250000  11229.000000   8.030310e+07        36.0     61.000000   
max    496987.000000  12221.000000   6.534642e+08        36.0     85.000000   

          tract_code    block_code  
count   55459.000000  55459.000000  
mean    39279.303522   2274.015110  
std     31409.179270   1309.307262  
min       100.000000   1000.000000  
25%     18101.000000   1003.000000  
50%     29300.000000   2001.000000  
75%     47500.000000   3001.000000  
max    162100.000000   9001.000000

Note that 5,329 out of the 60,788 rows have null values for `tract_code` (~9%) and therefore will not be counted in our processed data. We wish to understand if this is reflective of the household distribution or whether there is potential of bias with our processed data.

In [12]:
df_raw_nyc[df_raw_nyc.tract_code.isna()].groupby(['BOROUGH']).count()[['COURT_INDEX_NUMBER']]

COURT_INDEX_NUMBER
BOROUGH                          
BRONX                         898
BROOKLYN                      893
MANHATTAN                    1056
QUEENS                       2385
STATEN ISLAND                  97

Once again, Queens addresses as disproportionately impacted by the lack of `tract_code` value (45% compared to it's household percentage of 25%). Proportionally, Brooklyn is the least impacted (accounting for only 17% of the missing values whilst making up 30% of the number of households).

Further exploration is required to understand if the format of the address is incorrect for the API or whether the API is not complete and another should be used.